functorch/_src/vmap.py

```diff
@@ -355,15 +357,86 @@ def vmap(
         vmap_level = _vmap_increment_nesting(batch_size, randomness)
         try:
             batched_inputs = _create_batched_inputs(flat_in_dims, flat_args, vmap_level, args_spec)
             batched_outputs = func(*batched_inputs, **kwargs)
             return _unwrap_batched(batched_outputs, out_dims, vmap_level, batch_size, func)
         finally:
             _vmap_decrement_nesting()
-    return wrapped
+
+    def _get_chunk_flat_args(flat_args_, flat_in_dims_, chunks_):
+        flat_args_chunks = tuple(
+            t.chunk(chunks_, dim=in_dim) if in_dim is not None else [t, ] * chunks_
+            for t, in_dim in zip(flat_args_, flat_in_dims_)
+        )
+        # transpose chunk dim and flatten structure
+        # chunks_flat_args is a list of flatten args
+        chunks_flat_args = zip(*flat_args_chunks)
+        return chunks_flat_args
+
+    def _flatten_chunks_output(chunks_output_):
+        # chunks_output is a list of chunked outputs
+        # flatten chunked outputs:
+        flat_chunks_output = []
+        arg_spec_list = []
+        for output in chunks_output_:
+            flat_output, arg_specs = tree_flatten(output)
+            flat_chunks_output.append(flat_output)
+            arg_spec_list.append(arg_specs)
+
+        arg_spec = arg_spec_list[0]  # all specs should be the same
+        # transpose chunk dim and flatten structure
+        # flat_output_chunks is flat list of chunks
+        flat_output_chunks = list(zip(*flat_chunks_output))
+        return flat_output_chunks, arg_spec
+
+    @functools.wraps(func)
+    def wrapped_with_chunks(*args, **kwargs):
+        _check_out_dims_is_int_or_int_pytree(out_dims, func)
+        _, flat_in_dims, flat_args, args_spec = _process_batched_inputs(in_dims, args, func)
+        # Chunk flat arguments
+        chunks_flat_args = _get_chunk_flat_args(flat_args, flat_in_dims, chunks)
+
+        # Apply vmap on chunks
+        chunks_output = []
+        for flat_args in chunks_flat_args:
+            batch_size = _validate_and_get_batch_size(flat_in_dims, flat_args)
+            chunks_output.append(
+                _flat_vmap(batch_size, flat_in_dims, flat_args, args_spec, **kwargs)
+            )
+
+        flat_output_chunks, arg_spec = _flatten_chunks_output(chunks_output)
+        # Removing temporary variables helps to reduce memory usage on device like CUDA
+        del chunks_output
+
+        # concat chunks on out_dim
+        flat_out_dims = _broadcast_to_and_flatten(out_dims, arg_spec)
+        assert len(flat_out_dims) == len(flat_output_chunks)
+        flat_output = []
+        for out_dim in flat_out_dims:
+            flat_output.append(torch.cat(flat_output_chunks[0], dim=out_dim))
+            # release source data
+            del flat_output_chunks[0]
+        del flat_output_chunks
+
+        # finally unflatten the output
+        return tree_unflatten(flat_output, arg_spec)
+
+    return wrapped_with_chunks
```

Comment:
I didn't read very closely, but the approach looks good to me (and I will give a closer read after we figure out the other details).

To summarize what I think is going on here: vmap(..., chunks=N) does the following:
- chunks the inputs
- does vmap on each chunk to produce an output
- cats each output chunk to produce the original
and does not do anything special, like device offloading, right?

If so, I want to bikeshed the API a little more. I'm worried about two cases:
- It seems to me that people will end up calling library code (that uses vmap with chunks=1), but it turns out that the library code OOMs
- People call library code that calls vmap with chunks=5, because this was the most optimal number at the time. However, it's been 2 years and they now have a larger machine...

Then, there are two options:
1) the library code must thread a `chunks` argument around. The API that the user call has a `chunks` argument.
2) We provide a context-manager-based interface for chunks that can override the previously set chunk count

I don't know which one of these is better, but here's what I mean for Option 2.

## Proposal

We provide a `vmap_chunks(chunks=1, override=True)` context manager

The way to use it is:
```
with vmap_chunks(chunks=4):
  functorch.vmap(f)(x)

with vmap_chunks(chunks=4):
  functorch.hessian(f)(x)
```

The context managers are nestable. 
```
# The innermost one gets precedence
with vmap_chunks(chunks=4):
  with vmap_chunks(chunks=2):
    # uses 2 chunks
    functorch.hessian(f)(x)

# Unless the outer one specifies override
with vmap_chunks(chunks=4, override=True):
  with vmap_chunks(chunks=2):
    # uses 4 chunks
    functorch.hessian(f)(x)
```

What are your thoughts? @vfdev-5 , (also @samdow -- you've been working a lot with context managers recently)

Reply:
Thanks for the review @zou3519 !

As for the two options, I was thinking about the first option:

> the library code must thread a chunks argument around. The API that the user call has a chunks argument.

which means that user specifies `chunks` at each call like : 
```python
chunks = 4
out1 = functorch.hessian(f, chunks=chunks)(x)
out2 = functorch.jacrev(f, chunks=chunks)(x)
```

If we would like to avoid specifying it on every call and yet without making complex context managers, we can think of a global variable:
```python
from functorch.globals import set_num_chunks, get_num_chunks

set_num_chunks(4)
print(get_num_chunks())  # 4

out1 = functorch.hessian(f)(x)
out2 = functorch.jacrev(f)(x)
```

I think introducing `vmap_chunks(chunks=1, override=True)` context manager may bring a lot of unnecessary complexity to handle (multiple CMs, override/do not override etc). Please fix me if I'm missing something important for the API.

Reply:
I see the problem. I agree with you that if it is a nested call, option 2 (`jacfwd(jacrev(f), chunks=4)`) is a way to compute with chunks.

> Now, let's assume we did option 2, using the global variable. Then what exactly does the following do? Is it option 1 or option 2?

```python
set_num_chunks(4)
print(get_num_chunks())  # 4
out1 = functorch.hessian(f)(x)
# under the hood:
# return jacfwd(jacrev(f), chunks=get_num_chunks())
```

Reply:
> Maybe 1 should not be solved using the chunks API?

There are few workarounds here. a) We can say in the docs about chunks that hessian does chunk on cols (`num_cols_in_jacobian`), if this does not pass memory size, user can express manually the hessian as `jacfwd(jacrev(f, chunks=num_rows_in_jacobian), chunks=num_cols_in_jacobian)` and tune the params such that it could fit the memory.
b) try/except in hessian with  `num_cols_in_jacobian` only option and if fails with OOM try `jacfwd(jacrev(f, chunks=chunks), chunks=chunks)`. I'm not fan of such automagic options.
c) exceptionally for hessian we can have 2 chunk args `num_rows_in_jacobian`, `num_cols_in_jacobian`.

In `jacrev` there is also an expensive `_construct_standard_basis_for` which may be tried to to be done differently...

> "I am running into OOM and I want to tune a some numbers until my program stops OOM-ing)"

This looks like "auto" batch size search for NN. This can be an option.

Reply:
Punting this discussion to the future, per offline discussion